### DISCUSSION TITLE
Possible to increase PWM frequency and still have analogWrite() work?

### DISCUSSION THREAD

**Discussion by angryoctopus** [SpenceKonde/megaTinyCore#648] [Q&A] [answered]
I'm driving a Solenoid and get an annoying whine from the PWM frequency, I was wondering if it's possible to push the PWM frequency out of the audible range (like 10Khz) without having to manually configure TCA0.
I tried just lowering the clock prescalar in setup() with:
  TCA0.SPLIT.CTRLA = TCA_SINGLE_CLKSEL_DIV4_gc | TCA_SINGLE_ENABLE_bm;
But this fails during verification.

**Comment by SpenceKonde** [comment, marked as answer]
That should work, what error do you get? 

TCA_SINGLE_CLKSEL_DIV4_gc | TCA_SINGLE_ENABLE_bm should be correct. 

That said, I'm going to guess you are using a MOSFET to switch the power to the solenoid.

Two problems:
First, 10 kHz is still at an audible pitch, just barely, you'd want to run it faster than that; dividing the clock by 2 would be more like it. 
Secondly, at either 10 or 20 kHz.... are you using a gate driver IC on that MOSFET? 

If not... 10-20 kHz depending on the MOSFET? Switching losses would be high (how high depends on the MOSFET and what voltage the chip is running at), and power dissipation in the FET will hence be high. How fast you can pwm a gate directly before you run into trouble. Since I sell a bunch of MOSFET boards, some with gate drivers, I wrote a considerable document on MOSFET sizing after getting sick of answering the same questions about them.

https://github.com/SpenceKonde/ProductInfo/blob/master/MOSFETs/Guide.md

**Reply by angryoctopus** [reply]
readwrite.py", line 102, in read_data_words
return self.datalink.ld16(self,address)
TypeError: ld16() takes 2 positional arguments but 3 were given
TypeError: ld16() takes 2 positional arguments but 3 were given

[========================================== ] 6/7

That is the error I get when trying to upload.

I realized it should probably be:
TCA0.SPLIT.CTRLA = TCA_SPLIT_CLKSEL_DIV4_gc | TCA_SPLIT_ENABLE_bm;

But I still get the same error.

I am driving with a MOSFET (Currently a IRFZ44N) at 24V via an opto-coupler.
I'll definitely give that guide a read, looks like some great info!

**Reply by angryoctopus** [reply]
So after fixing the upload issue this works fine and the TCA_SPLIT_CLKSEL_DIV4_gc makes the switching inaudible to me and my Mosfet isn't warm to the touch (I assume opto-coupler driving the gate prevents any switching issues). Many thanks!

**Comment by SpenceKonde** [comment]
Thanks for reporting - that would be hitherto unreported bug in SerialUPDI. Looks like it would happen during verification when the final page has a single word in it; I don't know what I'm doing in python and made the same mistake in a few places. 

doing *anything* that changes the length of the compiled sketch should make the error go away until I have the fix in a release, it's tracked in #649 and will be fixed momentarily, it's a trivial fix. 

Glad to hear it - depending on the type of opto, the specs on those things can be profoundly disappointing w/rt switching speed. A while back I was looking to make an opto-isolated mosfet board for people who needed to do high-side switching . When I finally figured out how to interpret the specs regarding the switching speed of those things.... well... I still have a bag of parts from before I'd figured that out, and I don't have a P-channel FET board. They.... it's one of the cases where the specs seem to be intentionally obtuse because they' don't flatter the part. (just like how capacitor headline specs give a capacitance at 0v DC bias, and the maximum voltage, but don't say nothing about what the capacitance is near that voltage (some manufacturers will give you the disappointing graph on their website, a few even let you search on it, but many don't) and how battery datasheets give mAh at the precise load that maximizes the battery life, and MOSFET manufacturers tell you the maximum continuous drain current at "Tc = 25C")

**Reply by angryoctopus** [reply]
You are correct, I was wracking my brain as to how a code change that compiles fine could stop the upload/validation from working.

**Reply by angryoctopus** [reply]
Ironically I was intending to switch to a P-Channel mosfet because I have some boards arriving with the solenoid accidently on the source side and I figured the rework would be a lot easier by just switching to a P-Channel and moving the gate.
I'm using PC817 optocouplers and I'll just have to see how it works, hopefully won't have to deep dive into datasheets.
